### PR TITLE
Remove the scheduling decision based on PreviousAssignment

### LINKED ISSUE #984: TODO: Scheduling the tasks without considering PreviousAssignment
It is better to avoid having multiple sources of truth while making decisions on the tasks that have been already scheduled. Hence, it is better to rely in CurrentState completely and get rid of PreviousAssignment.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #984 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In this commit, several parts of the code have been removed in order to avoid
making scheduling decision based on PreviousAssignment. Instead we are relying on
the CurrentState. 
In order to facilitate this, the tasks that reaches the terminal state 
(Error, Time_out, Aborted and completed) will be dropped and rescheduled in later
pipeline if needed (If they need to be rescheduled). This helps us to reduce the
number of existing current states for each task. 

### Tests
- [x] The following is the result of the "mvn test" command on the appropriate module:
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestWorkflowTermination.testWorkflowJobFail:251->verifyWorkflowCleanup:257 expected:<true> but was:<false>
[ERROR]   TestWorkflowTimeout.testWorkflowTimeoutWhenWorkflowCompleted:116 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 1144, Failures: 2, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:21 h
[INFO] Finished at: 2020-05-01T15:25:20-07:00
[INFO] ------------------------------------------------------------------------

The failed tests has passed when I ran it individually. 

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 1]
```diff
@@ -99,16 +99,15 @@ public void updatePreviousAssignedTasksStatus(
...
 
       // We need to remove all task pId's to be dropped because we already made an assignment in
       // paMap above for them to be dropped. The following does this.
       if (tasksToDrop.containsKey(instance)) {
         pSet.removeAll(tasksToDrop.get(instance));
       }
 
-      // Used to keep track of partitions that are in one of the final states: COMPLETED, TIMED_OUT,
-      // TASK_ERROR, ERROR.
+      // Used to keep track of partitions that are in one the INIT or DROPPED states.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:110]: Did you mean "that are in either INIT or DROPPED states"?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 2]
```diff
@@ -242,16 +230,16 @@ public void updatePreviousAssignedTasksStatus(
         }
           break;
         case COMPLETED: {
-          // The task has completed on this partition. Mark as such in the context object.
-          donePartitions.add(pId);
+          // The task has completed on this partition. Drop it from the instance and add it to assignedPartitions in
+          // order to avoid scheduling it again in this pipeline.
+          assignedPartitions.get(instance).add(pId);
+          paMap.put(pId, new PartitionAssignment(instance, TaskPartitionState.DROPPED.name()));
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:236]: I don't know if this would be the correct thing to do.

Sure, once a task is complete, it could be dropped from the node, but you also want to keep the state of the task in COMPLETED in JobContext.

I'm afraid doing this might cause the otherwise-completed task's state to change to DROPPED. Then we would have no way of knowing if this task just got dropped or completed and dropped.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 3]
```diff
@@ -263,7 +251,11 @@ public void updatePreviousAssignedTasksStatus(
         case TASK_ABORTED:
 
         case ERROR: {
-          donePartitions.add(pId); // The task may be rescheduled on a different instance.
+          // First make this task which is in terminal state to be dropped.
+          // Later on, in next pipeline in handleAdditionalAssignments, the task will be retried if possible.
+          // (meaning it is not ABORTED and max number of attempts has not been reached yet)
+          assignedPartitions.get(instance).add(pId);
+          paMap.put(pId, new PartitionAssignment(instance, TaskPartitionState.DROPPED.name()));
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:258]: Again, I am not sure if this would cause the JobContext to have the wrong state for the task partitions. Did you check? What state does the task go into between being dropped and getting rescheduled? (It should be ERROR or TASK_ABORTED, or whatever it was before being dropped)

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 4]
```diff
@@ -389,13 +381,57 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
+   * @param currentState
+   * @param jobCtx
+   * @param pId
+   * @param pName
+   * @param instance
+   */
+  private void updatePartitionInformationInJobContext(CurrentStateOutput currentStateOutput,
+      String jobResource, TaskPartitionState currentState, JobContext jobCtx, Integer pId,
+      String pName, String instance) {
+    // The assigned participant needs to be updated regardless of the current state and context
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:404]: the assignedParticipant "field" in JobContext needs to be updated...

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 5]
```diff
@@ -389,13 +381,57 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
+    // information because it will prevent controller to stuck in race condition while there is two
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:405]: ... because this will prevent the controller from being stuck in a race condition when there are two current states found for a single task.

-> Okay, but what kind of race condition are we talking about?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 6]
```diff
@@ -389,13 +381,57 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
+    // current states. In the updatePreviousAssignedTasksStatus, we check
+    // instance.equals(jobCtx.getAssignedParticipant(pId)) and bypass the assignment if instance is
+    // not equal to job context's AssignedParticipant for this pId.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:408]: Does this help us avoid the race condition?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 7]
```diff
@@ -389,13 +381,57 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
     jobCtx.setAssignedParticipant(pId, instance);
-    jobCtx.setPartitionState(pId, currentState);
-    String taskMsg = currentStateOutput.getInfo(jobResource, new Partition(pName), instance);
-    if (taskMsg != null) {
-      jobCtx.setPartitionInfo(pId, taskMsg);
+    // If job context needs to be updated with new state, update it accordingly
+    // This check is necessary because we are relying on current state and we do not want to update
+    // context as long as current state existed. We just want to update context information
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:412]: -> update context while its current state exists...

But why do we not want to update the context when there's a current state for it? Shouldn't we update the context based on the latest current state?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 8]
```diff
@@ -389,13 +381,57 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
+    // (specially finish time) once.
+    if (!currentState.equals(jobCtx.getPartitionState(pId))) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:414]: This condition checks whether jobContext's state is out of date or not, right?

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java [hunk 9]
```diff
@@ -380,72 +379,38 @@ private static boolean isJobComplete(JobContext ctx, Set<Integer> allPartitions,
...
-
-    // Add all pIds existing in CurrentStateOutput as well because task currentStates copied over
-    // from previous sessions won't show up in prevInstanceToTaskAssignments
-    // We need to add these back here in order for these task partitions to be dropped (after a
-    // copy-over, the Controller will send a message to drop the state currentState)
-    // partitions: (partition -> instance -> currentState)
+    // Generate currentInstanceToTaskAssignment with CurrentStateOutput as source of truth
+    // Add all pIds existing in CurrentStateOutput
+    // We need to add these pIds here and make decision about the existing tasks with current state
+    // in updatePreviousAssignedTasksStatus method
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:399]: "add these pIds to `result`" and update their states in JobContext in `updatePreviousAssignedTasksStatus` method...

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java [hunk 10]
```diff
@@ -380,72 +379,38 @@ private static boolean isJobComplete(JobContext ctx, Set<Integer> allPartitions,
...
-          // ** for tasks that are just in INIT state, we do not add them here because old
-          // Participants, upon connection reset, set tasks' currentStates to INIT. We cannot
-          // consider those tasks active **
-          if (currState == TaskPartitionState.INIT && requestedState != null
-              && requestedState.equals(TaskPartitionState.RUNNING.name())
-              || currState == TaskPartitionState.RUNNING) {
-            result.get(instance).add(pId);
-          }
-
+          // We must add all pIds back here
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:410]: This comment isn't giving any information. Why must we?

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java [hunk 11]
```diff
@@ -380,72 +379,38 @@ private static boolean isJobComplete(JobContext ctx, Set<Integer> allPartitions,
...
+          result.get(instance).add(pId);
           // Check if this task needs to be dropped. If so, we need to add to tasksToDrop no matter
-          // what its current state is so that it will be dropped
+          // what its current state is so that it will be dropped.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:413]: Nit: This dot isn't necessary...
If anything, repeat again what "a task needs to be dropped" means. This is trying to drop tasks on a reconnected instance with a new sessionId that have all of their requestedState == DROPPED.

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java [hunk 12]
```diff
@@ -462,10 +427,10 @@ private static boolean isJobComplete(JobContext ctx, Set<Integer> allPartitions,
    * If partition is missing from prevInstanceToTaskAssignments (e.g. previous assignment is
    * deleted) it is added from context. Otherwise, the context won't be updated.
    * @param jobCtx Job Context
-   * @param prevInstanceToTaskAssignments instance -> partitionIds from previous assignment
+   * @param currentInstanceToTaskAssignments instance -> partitionIds from current assignment
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:430]: "from CurrentStateOutput"

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 13]
```diff
@@ -242,16 +230,16 @@ public void updatePreviousAssignedTasksStatus(
         }
           break;
         case COMPLETED: {
-          // The task has completed on this partition. Mark as such in the context object.
-          donePartitions.add(pId);
+          // The task has completed on this partition. Drop it from the instance and add it to assignedPartitions in
+          // order to avoid scheduling it again in this pipeline.
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:234]: Why you remove the done partition from COMPLETED state? Then this partition will not be removed from pSet, right?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:110]: Yes. Done.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 13 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:234]: The reason is the behavior has been changed in this PR. We have an assignment for it and the assignment is dropped. So, it goes to assigned partitions. Once it is dropped from participant, then we do not even consider it for the new assignment.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 2 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:236]: You are right. I also had this concern. But note that we do not change the JobContext to DROPPED. Let’s say controller sends COMPLETED to DROPPED message to the participant. Now, the current state will be COMPLETED until participant processes the message. Once it is processed, the current state will be null, and that task will not be considered again. So, to answer your question, it is totally depending on the behavior of the participant. If participant set the current state  to DROPPED, which is not the case here, your concern was correct. Since participant removes the current state of the task, we do not need to be worry about that case.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 3 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:258]: The answer to this question is also similar to COMPLETED case :). Please see it above.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 4 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:404]: Fixed. Thanks.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 5 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:405]: The race condition that here is scenario that I am talking about is the scenario where we have two current states of a task in different participants. Since we have only one slot for each task in paMap, we want to make sure that the task is DROPPED from the wrong one and is assigned to the correct instance. I am basically talking about this #922 and #461. Let me rephrase this comment to make it clear.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 6 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:408]: Exactly. We add this check in PR #923 to help resolve this scenario. Maybe using race condition is not the good word here. I rephrased the comments.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 8 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:414]: Exactly. I added a comment about it.

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java (hunk 9 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:399]: Fixed. Thanks.

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java (hunk 10 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:410]: I agree. Removed the comment. Basically, we have explained everything in the comments already.

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java (hunk 11 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:413]: Done.

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java (hunk 12 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:430]: Done.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 7 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:412]: Here we check the if current state has been changed and it is not equal to context information. This check is necessary because we do not need to update the context in every pipeline. Also, the more important part of this check is for the fields such as finish time. Let’s say we have a delay retry time for the task. This functionality depends on the finish time in the context. If we update finish time as long as we have current state (note that we set it to current time), we will never schedule the task with delay retry time. Because finish time in the context and delay retry time will be messed up. That is why we only update the context at the first time that the current state has been changed.

**Comment by alirezazamani** [comment]
> I have a general suggestion. Let's do the renaming in another PR. Then we can have clear understand what is the logical change.

There are not that much name changes. It is basically prevAssignment to currentState. The rest are similar. If you still think the name change is necessary, I can change the names back to original. Please let me know.

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java [hunk 14]
```diff
@@ -238,21 +238,20 @@ private ResourceAssignment computeResourceMapping(String jobResource,
...
 
     if (LOG.isDebugEnabled()) {
       LOG.debug("All partitions: " + allPartitions + " taskAssignment: "
-          + prevInstanceToTaskAssignments + " excludedInstances: " + excludedInstances);
+          + currentInstanceToTaskAssignments + " excludedInstances: " + excludedInstances);
     }
 
     // Release resource for tasks in terminal state
-    updatePreviousAssignedTasksStatus(prevInstanceToTaskAssignments, excludedInstances, jobResource,
+    updatePreviousAssignedTasksStatus(currentInstanceToTaskAssignments, excludedInstances, jobResource,
```
Comment (NealSun96) [inline helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:254]: Should this function be renamed?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 15]
```diff
@@ -121,17 +120,6 @@ public void updatePreviousAssignedTasksStatus(
               instance, pId);
           continue;
         }
```
Comment (NealSun96) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:122]: Would it be a problem that this skipping check is now done after markPartitionCompleted/markPartitionError?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 16]
```diff
@@ -389,13 +381,59 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
+    // This check is necessary because we are relying on current state and we do not want to update
+    // context as long as current state existed. We just want to update context information
+    // (specially finish time) once.
+    // This condition checks whether jobContext's state is out of date or not.
+    if (!currentState.equals(jobCtx.getPartitionState(pId))) {
+      jobCtx.setPartitionState(pId, currentState);
+      String taskMsg = currentStateOutput.getInfo(jobResource, new Partition(pName), instance);
+      if (taskMsg != null) {
+        jobCtx.setPartitionInfo(pId, taskMsg);
+      }
```
Comment (NealSun96) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:421]: Is there a possibility that a message exists without state changes?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 13 above)
Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:234]: The concern for doing this is that will this reflect the change to JobContext? So if final state will all be DROPPED. It is hard for us to determine whether the tasks is completed or it requires another retry? Because if task has problem in the middle of running, it will be DROPPED as well.

Comment (dasahcc) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:234]: Also, I dont think you will get DROPPED state as current state, right? Because the current state will be deleted from the current state if the to state is DROPPED after it completes state transition.

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:234]: Exactly. Since we do not get DROPPED state in current state, we will not mark the the task as DROPPED in job context. If the task is completed and we send COMPLETE to DROPPED message, then the participant will DROP the currentstate and this function will not see the task on the instance again because we are based on CurrentState now. Hence, we will not mark the task DROPPED in jobContext. Basically since in this function context follows the currentState, then DROPPED context will not be possible. Right?

File: helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java (hunk 14 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java:254]: I think it should be fine to keep it as it is. Because we are updating the state of the tasks that has been assigned in previous pipelines based on the currentstate of the tasks.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 15 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:122]: This shouldn't be an issue because this check is specifically added before to bypass the scenario where we have two current states which non of the markPartitionCompleted/markPartitionError will not be hit because it is either Running or Dropped. Good question though.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 16 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:421]: The participant read the message, handle state change and then delete the message. So when messages get deleted from ZK after ST is completed. So we should be safe here.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 7 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:413]: @alirezazamani -

Okay, I think your point about not updating when there's no delta is valid, but I'm not sure if what you said about delayed scheduling is entirely true. Delayed scheduling of tasks is a feature that was previously working, right? If what you said about it was the case, then how come our delayed scheduling feature was working all this time?

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java [hunk 17]
```diff
@@ -389,13 +381,59 @@ private TaskPartitionState updateJobContextAndGetTaskCurrentState(
...
-    jobCtx.setPartitionState(pId, currentState);
-    String taskMsg = currentStateOutput.getInfo(jobResource, new Partition(pName), instance);
-    if (taskMsg != null) {
-      jobCtx.setPartitionInfo(pId, taskMsg);
+    // If job context needs to be updated with new state, update it accordingly
+    // This check is necessary because we are relying on current state and we do not want to update
+    // context as long as current state existed. We just want to update context information
+    // (specially finish time) once.
+    // This condition checks whether jobContext's state is out of date or not.
+    if (!currentState.equals(jobCtx.getPartitionState(pId))) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:416]: @alirezazamani Another concern here:

Is it possible for a task to be retried so fast that it ends up being in the same state? For example,

task_error -> (controller reschedules it) -> (controller sends messages, error -> init, init -> running, running -> complete) -> participant processes message so quickly but it goes into task_error again, and by the time controller gets to this task, the states are the same as the previous run. Basically, controller sees task_error ->  task_error.

Is that going to be an issue? I guess partition id is set before this check so that will be updated accordingly, but what about finish time or the time at which the task was marked as error?

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:416]: @narendly Thank you for mentioning these corner cases. 

I went through the logic and I am sure the code handles this corner case as well. Let me explain why (please note that if task goes to terminal states, with this PR we are dropping 
 the task first and schedule them in next pipeline):

Here are the steps:
1- Requested State is set to TASK_ERROR by the participant. 
2- Controller sends RUNNING -> TASK_ERROR to participant.
3- Participant process it and makes current state as TASK_ERROR.
4- Controller see this (i.e. TASK_ERROR) current state and send TASK_ERROR -> DROPPED message.
5- Participant gets the message and drops/removes current states.
6- Controller does not see current state (because it is removed) and sets the context of the task to be INIT.  Schedule the task again on the participant by sending INIT-> RUNNING message. (this happens in handleAdditionalTaskAssignment method).
*Now controller by itself does not send RUNNING -> COMPLETE because COMPLETE state should be requested by the participant. The participant will either request TASK_ERROR or COMPLETE state in current state.
7- If the task goes to error state we see this delta again (because context is set to be INIT before scheduling the task) and mark it ERROR with new finish time. 

There are two points here that helps controller to handle this cornet case:
1- controller will not send running -> complete unless participant requested state is COMPLETE. So if the task goes to TASK_ERROR in participant, the requested state will be TASK_ERROR and controller always respects the requested state and send RUNNING -> TASK_ERROR. 
2- Dropping the task in terminal state makes the controller to reschedule the task with jobContext as INIT. This behavior helps the controller to see these deltas in future pipelines and update the information correctly in jobContext.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 7 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:413]: @narendly 

The reason that delayed scheduling was working is because we were relying on previousAssignment. Let's say we are relying on previousAssignment and we decided to send the task to TASK_ERROR. This task will be recorded in previousAssignment and next pipeline we iterate through the task existed in previous assignment and record finish time and context information. Then this task will not be existed in future pipelines previousAssignment and we do not update finish time in the future pipelines. (Actually some of the test guided me toward this change).

However, if we want to be independent of previousAssignment and rely on currentState (which is also means we are relying on participant's reactions), we need to consider that currentState might not change for several pipelines. In this case, it is necessary to monitor these deltas and not update the context information (specially time sensitive information) multiple times.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 17 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:416]: > 1- controller will not send running -> complete unless participant requested state is COMPLETE

Are you sure this is true? Can you verify that init -> running / running -> complete are not sent out at the same time? The target state during message generation phase should be complete if I recall correctly.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 7 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:413]: Thanks for the explanation. The reason we need to know if CurrentState changed is because we no longer have prevAssignment.

File: helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java (hunk 17 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:416]: I double checked again and it seems the only way that controller sends COMPLETED message to participant should be initiated by participant by setting the requested state to be completed.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/AbstractTaskDispatcher.java:416]: Sounds good. Overall I think this analysis is sound. Let's make sure we do a careful round of  E2E testing as well as load testing.

**Comment by alirezazamani** [comment]
mvn test result before merge:
Helix-Core
```
[INFO] Tests run: 1144, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4,939.921 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1144, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:22 h
[INFO] Finished at: 2020-05-26T10:42:39-07:00
[INFO] ------------------------------------------------------------------------
```
Helix-rest:
```
[INFO] Tests run: 159, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 46.025 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 159, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  51.874 s
[INFO] Finished at: 2020-05-26T16:31:12-07:00
[INFO] ------------------------------------------------------------------------
```

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @narendly 

Final commit message:

Make the task scheduling decision independent of the PreviousAssignment

In this commit, the previous logic which was based on PreviousAssignment, has been changed. Instead, the task scheduling is done based on the CurrentState.
